torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -300,7 +324,14 @@ class ProcessGroupNCCL : public ProcessGroup {
     c10::intrusive_ptr<Future> then(
         std::function<at::IValue(void)> callback,
         at::TypePtr /* unused */) override {
-      auto fut = c10::make_intrusive<FutureNCCL>(deviceIndex_);
+      auto fut = c10::make_intrusive<FutureNCCL>();
+      // The new future needs the DataPtr extractor when it gets marked complete
+      // but this might happen immediately inline or in parallel by another
+      // thread. In both these cases this would/might happen before the user has
+      // time to set their own DataPtr extractor, which might lead to failures
+      // if the default extractor can't handle some of the user's types.
+      // Therefore we propagate our extractor.
```

Comment:
I realized that since we now use the DataPtr extractor also in markCompleted, we need to do this ugly thing. I suspect that this was the cause of the CI failures (we'll see shortly). I welcome ideas on how to do this better...